src/librustc_mir/borrow_check/diagnostics/mod.rs

```diff
@@ -772,7 +776,13 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             }
         }
 
-        let normal_ret = OtherUse(stmt.source_info.span);
+        let normal_ret = if let [ProjectionElem::Downcast(..), ProjectionElem::Field(_, _)] =
+            moved_place.projection
+        {
+            PatUse(stmt.source_info.span)
+        } else {
+            OtherUse(stmt.source_info.span)
+        };
```

Comment:
I'm really unsure about this, but worked as a placeholder way of accomplishing my goal. If there's a more reliable way of detecting that the move occurs in a pattern, I can change this.

Reply:
The correct (or at least more correct) way to check if this is a pattern binding is to check that:

- This location is an `StatementKind::Assign` of an `Rvalue::Use`
- The assignment is to a user defined local (i.e. `self.body.local_decls[target].is_user_variable()` is true
- The ever initialized flow is false at this point.

Unfortunately the flows aren't available here currently, and changing things so that they are would be a fairly large change (#73552).

I think that if this is going to use an imperfect heuristic that would be fine with a fixme referencing that issue. The condition should be `moved_place.projection.iter().any(|p| matches!(p, ProjectionElem::Downcast)`.

Reply:
Changed to use the imperfect heuristic.